src/GafferScene/SetExpression.cpp

```diff
@@ -0,0 +1,276 @@
+
+using namespace Gaffer;
+using namespace GafferScene;
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+
+#include <boost/config/warning_disable.hpp>
+#include <boost/spirit/include/qi.hpp>
+#include <boost/variant/recursive_variant.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/spirit/include/phoenix_operator.hpp>
+#include <boost/spirit/include/phoenix_function.hpp>
+
+#include <iostream>
+#include <vector>
+#include <string>
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+
+// AST
+struct binary_op;
+struct nil {};
+
+struct ExpressionAst
+{
+	typedef
+	boost::variant<
+		nil, // can't happen!
+		std::string,
+		boost::recursive_wrapper<ExpressionAst>,
+		boost::recursive_wrapper<binary_op>
+		>
+	type;
+
+	ExpressionAst()
+		: expr(nil()) {}
+
+	template <typename Expr>
+	ExpressionAst(Expr const& expr)
+		: expr(expr) {}
+
+	ExpressionAst& operator+=(ExpressionAst const& rhs);
+	ExpressionAst& operator-=(ExpressionAst const& rhs);
+
+	type expr;
+};
+struct binary_op
+{
+	binary_op(
+		char op,
+		ExpressionAst const& left,
+		ExpressionAst const& right)
+		: op(op), left(left), right(right) {}
+
+	char op;
+	ExpressionAst left;
+	ExpressionAst right;
+};
+
+ExpressionAst& ExpressionAst::operator+=(ExpressionAst const& rhs)
+{
+	expr = binary_op('|', expr, rhs);
+	return *this;
+}
+
+ExpressionAst& ExpressionAst::operator-=(ExpressionAst const& rhs)
+{
+	expr = binary_op('&', expr, rhs);
+	return *this;
+}
+
+// Visiting the AST
+// ----------------
+struct AstPrinter
+{
+typedef void result_type;
+
+void operator()(qi::info::nil) const {}
+void operator()(std::string n) const { std::cout << n; }
+
+void operator()(ExpressionAst const& ast) const
+{
+	boost::apply_visitor(*this, ast.expr);
+}
+
+void operator()(binary_op const& expr) const
+{
+	std::cout << "op:" << expr.op << "(";
+	boost::apply_visitor(*this, expr.left.expr);
+	std::cout << ", ";
+	boost::apply_visitor(*this, expr.right.expr);
+	std::cout << ')';
+}
+
+};
+
+struct AstEvaluator
+{
+typedef PathMatcher result_type;
+
+AstEvaluator( const ScenePlug *scene )
+{
+	m_scene = scene;
+}
+
+result_type operator()(std::string setName) const
+{
+	// TODO: handle case in which the setName given is not a valid set
+	return m_scene->set( setName )->readable();
+}
+
+result_type operator()(ExpressionAst const& ast) const
+{
+	return boost::apply_visitor(*this, ast.expr);
+}
+
+result_type operator()(binary_op const& expr) const
+{
+	PathMatcher left = boost::apply_visitor(*this, expr.left.expr);
+	PathMatcher right = boost::apply_visitor(*this, expr.right.expr);
+
+	if( expr.op == '|' )
+	{
+		PathMatcher result = PathMatcher( left );
+		result.addPaths( right );
+		return result;
+	}
+	else if( expr.op == '&' )
+	{
+		PathMatcher result = PathMatcher();
+
+		for(PathMatcher::Iterator it = left.begin(); it != left.end(); ++it)
+		{
+			if( right.find( *it ) != right.end() )
```

Comment:
I'm not sure I follow. I had tried the `RawInterator` first as that is what find() returns, but it gave me exactly the problem you describe: it matched all kinds of ancestor locations. When I switched to the current implementation, which should only visit the actual locations, that problem went away.

I just did another test using the added `GafferSceneTest.PathMatcherTest.PathMatcherTest.testIntersection()` unittest and checked both `Interator` and `RawInterator`. Iterator works fine while `RawIterator` gives me this - which I kinda expected:

```
Traceback (most recent call last):
  File "/home/mattig/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferSceneTest/PathMatcherTest.py", line 1011, in testIntersection
    self.assertEqual( m3.paths(), [ "/a/b/c/d/myTest" ] )
AssertionError: Lists differ: ['/', '/a', '/a/b', '/a/b/c', ... != ['/a/b/c/d/myTest']

First differing element 0:
/
/a/b/c/d/myTest

First list contains 5 additional elements.
First extra element 1:
/a

- ['/', '/a', '/a/b', '/a/b/c', '/a/b/c/d', '/a/b/c/d/myTest']
+ ['/a/b/c/d/myTest']
```

Maybe I'm missing something, but I had thought that using `Iterator` instead of `RawIterator` will just magically fix that issue..?